iroh-net/src/net/udp.rs

```diff
@@ -0,0 +1,150 @@
+        Self::bind(Network::Ipv6, port).await
+    }
+
+    /// Bind only Ipv4 on localhost.
+    pub async fn bind_local_v4(port: u16) -> Result<Self> {
+        Self::bind(Network::Ipv4, port).await
+    }
+
+    /// Bind only Ipv6 on localhost.
+    pub async fn bind_local_v6(port: u16) -> Result<Self> {
+        Self::bind(Network::Ipv6, port).await
+    }
+
+    /// Bind to the given port only on localhost.
+    pub async fn bind_local(network: Network, port: u16) -> Result<Self> {
+        let addr = SocketAddr::new(network.local_addr(), port);
+        Self::bind_raw(addr, true)
+            .await
+            .with_context(|| format!("{addr:?}"))
+    }
+
+    /// Bind to the given port and listen on all interfaces.
+    pub async fn bind(network: Network, port: u16) -> Result<Self> {
+        let addr = SocketAddr::new(network.default_addr(), port);
+        Self::bind_raw(addr, true)
+            .await
+            .with_context(|| format!("{addr:?}"))
+    }
+
+    /// Bind to any provided [`SocketAddr`]. Does not prepare for using the socket as QUIC socket.
+    pub async fn bind_full(addr: impl Into<SocketAddr>) -> Result<Self> {
+        Self::bind_raw(addr, false).await
+    }
+
+    async fn bind_raw(addr: impl Into<SocketAddr>, prepare_for_quinn: bool) -> Result<Self> {
+        let addr = addr.into();
+        let network = Network::from(addr.ip());
+        let socket = socket2::Socket::new(
+            network.into(),
+            socket2::Type::DGRAM,
+            Some(socket2::Protocol::UDP),
+        )?;
+
+        if let Err(err) = socket.set_recv_buffer_size(SOCKET_BUFFER_SIZE) {
+            warn!(
+                "failed to set recv_buffer_size to {}: {:?}",
+                SOCKET_BUFFER_SIZE, err
+            );
+        }
+        if let Err(err) = socket.set_send_buffer_size(SOCKET_BUFFER_SIZE) {
+            warn!(
+                "failed to set send_buffer_size to {}: {:?}",
+                SOCKET_BUFFER_SIZE, err
+            );
+        }
+        if network == Network::Ipv6 {
+            // Avoid dualstack
+            socket.set_only_v6(true)?;
+        }
+
+        if prepare_for_quinn {
+            quinn_udp::UdpSocketState::configure((&socket).into())?;
+            // disable nonblocking to ensure socket2 bind works
+            socket.set_nonblocking(false)?;
+        }
+
+        socket.bind(&addr.into())?;
+
+        // Reenable nonblocking
+        socket.set_nonblocking(true)?;
+
+        let socket: std::net::UdpSocket = socket.into();
+
+        // Convert into tokio UdpSocket
+        let socket = tokio::net::UdpSocket::from_std(socket)?;
+
+        if addr.port() != 0 {
+            let local_addr = socket.local_addr()?;
+            ensure!(
+                local_addr.port() == addr.port(),
+                "wrong port bound: {:?}: wanted: {} got {}",
+                network,
+                addr.port(),
+                local_addr.port(),
+            );
```

Comment:
In what situations does this assertion ever fail?  I mean, I would never have thought of writing such an assertion and am horrified this might be needed.

Reply:
I had a lot of fun debugging network bindings not doing what I expect, and I agree this should never happen..but my paranoia is strong at this point..